### ISSUE TITLE
Decorrelate subqueries with ORDER BY / TopN

### ISSUE THREAD

**Issue by findepi** [prestosql/presto#1358]
Example

```sql
SELECT a.*,
       (SELECT b.rating FROM b
        WHERE b.employee_id = a.employee_id
        ORDER BY b.date DESC LIMIT 1) as rating
FROM a;
```

**Comment by findepi** [comment]
I assume `LIMIT 1` will be a common factor in this type of queries.
We should implement this by automatically converting `SELECT x ... ORDER BY y LIMIT 1` into `SELECT min_by(x, y).

**Comment by kasiafi** [comment]
`max_by(x, y)` and `min_by(x, y)` ignore all rows where `y` is `null`.
So, in the case where the null ordering `NULLS FIRST` is specified (or implicit), it might return a different value than `TopN` would.
Moreover,  if all `y` values are `null`, it will return `null` instead of any `x` value`.

**Comment by kokosing** [comment]
> I assume `LIMIT 1` will be a common factor in this type of queries.

Any other LIMIT could be simply removed from the query.

**Comment by sopel39** [comment]
> Any other LIMIT could be simply removed from the query.

Maybe we can rewrite correlated `ORDER BY col LIMIT n` to:
* grouped `TopN` (by correlation columns) in subquery. This will lead to having `n` top rows for every matching correlation.
* we also need to apply grouped `TopN` above decorrelated join in order to restore order (join build rows might be produced in random order).

**Comment by sopel39** [comment]
For reference, grouped `TopN` is `TopNRowNumberNode``

**Comment by kasiafi** [comment]
I don't think we can decorrelate a subquery containing TopN.
However, TopN can be removed from a subquery and re-applied on top of decorrelated join.

It could take the form of TopNRowNumberNode partitioned by a unique row id which in this case would have to be added to the left join source.

**Comment by kasiafi** [comment]
Additionally, we need to make sure that:
- any Project in a subquery is restored above (plus remove row_number symbol, plus identities on left source symbols),
- TopN's ordering scheme is not correlated,
- TopN's count is 1, in the case when EnforeceSingleRow is present.
- LateralJoin type is suitable for this transformation.

**Comment by sopel39** [comment]
> I don't think we can decorrelate a subquery containing TopN.
> However, TopN can be removed from a subquery and re-applied on top of decorrelated join.

If limit is 1, then you can just rewrite `TopN` into `TopNRowNumberNode` with correlation symbols as partitioning symbols

If limit is >1, you also need to re-apply `TopNRowNumberNode` on top of de-correlated join.

> TopN's ordering scheme is not correlated,

Technically, correlated symbols are constants from subquery perspective. You could just rewrite `TopN` (with limit `N`) into `Limit N` (remove order).

> TopN's count is 1, in the case when EnforeceSingleRow is present.

That's not needed unless `TopN` physically produces more than 1 row

> any Project in a subquery is restored above (plus remove row_number symbol, plus identities on left source symbols),

I think you can extend `PlanNodeDecorrelator` with support for `TopN` (limit 1). For limit >1, you need extra transformation rule

**Comment by sopel39** [comment]
> Technically, correlated symbols are constants from subquery perspective. You could just rewrite TopN (with limit N) into Limit N (remove order).

Or you can do other way around. That is convert `Limit N` to `TopNRowNumberNode` on correlated symbols. This gives nice benefit of supporting `Limit N` in lateral join correlated subqueries!

**Comment by kasiafi** [comment]
Got it. `Limit` / `TopN` can indeed be transformed _in situ_, which is much better than pulling limit above join.

You said before that for `TopN` where N > 1, we need to reproduce partitioned sort above decorrelated join.
But does `LateralJoin` with sorted subquery grant any order on output? If it doesn’t, we don’t need to reproduce it.

**Comment by sopel39** [comment]
> But does LateralJoin with sorted subquery grant any order on output? If it doesn’t, we don’t need to reproduce it.

I'm not sure you can assume that. Although execution might produce ordered result, one should consult property derivation (see: `io.prestosql.sql.planner.optimizations.PropertyDerivations.Visitor#visitJoin`) to check if that's the case.

I this particular case I think it will be the case, but you would need to modify `io.prestosql.sql.planner.optimizations.PropertyDerivations.Visitor#visitJoin` to reflect that

**Comment by findepi** [comment]
There are no derived properties for LateralJoin:
https://github.com/prestosql/presto/blob/ab0382cdd912bb1c8e5b5a0f7ac3733b13b51ab4/presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java#L215-L219

> But does LateralJoin with sorted subquery grant any order on output? 

I'm sure it doesn't. See @martint 's https://prestosql.io/blog/2019/06/03/redundant-order-by.html quote from SQL specification (ISO 9075 Part 2):

> A `<query expression>` can contain an optional `<order by clause>`. The ordering of the rows of the table specified by the `<query expression>` is guaranteed only for the `<query expression>` that immediately contains the `<order by clause>`.

**Comment by kasiafi** [comment]
Right, so I assume that the ordering specified for the subquery's `TopN` shouldn't be expected, or referred to, in non-immediate context (above join).

**Comment by kasiafi** [comment]
> I this particular case I think it will be the case, but you would need to modify io.prestosql.sql.planner.optimizations.PropertyDerivations.Visitor#visitJoin to reflect that

`PropertyDerivations` currently guarantees keeping probe side order in left and inner join, in the case of no spill.
In our case, if the probe side order is kept, nothing can be assumed about the overall order of the decorrelated join's result, that is not captured by `PropertyDerivations`.

**Comment by kasiafi** [comment]
To sum up the topic of implementation:
each of the cases: `Top 1`, `Top N`, `Limit N` can be transformed in the same way, that is rewritten to `TopNRowNumberNode` in `PlanNodeDecorrelator`.
That will enable decorrelating not only scalar subqueries, but any `LateralJoin`’s subqueries containing `Limit` or `TopN`.

**Comment by sopel39** [comment]
> Limit N can be transformed in the same way, that is rewritten to TopNRowNumberNode in PlanNodeDecorrelator.

Technically, we could have a dedicated grouped limit operator. It could have a better performance. I would start with rewrite to `TopNRowNumberNode` though.

**Comment by sopel39** [comment]
> You said before that for TopN where N > 1, we need to reproduce partitioned sort above decorrelated join.

I don't think we need to reproduce sort above decorrelated join after all. I don't think SQL requires that. What would the order be anyway?
